src/syntax_util.ml

```diff
@@ -3,7 +3,32 @@ open Asttypes
 open Parsetree
 open Longident
 
-let a = Str.global_replace
+(** [is_prefixed prefix str] checks if prefix is the prefix of str *)
+let is_prefixed prefix str =
+  let len = String.length prefix in
+  if len > String.length str then
+    1
+  else
+    String.compare (String.sub str 0 len) prefix
+
+(** [replace_string old_str new_str str] replaces old_str to new_str in str *)
+let rec replace_string old_str new_str str =
```

Comment:
How long is `str` in practice here?
When `old_str` is not contained in `str`, for example, there will be a linear number of copies of (prefixes of) `str`, all in memory (not tail recursive), for quadratic time+space.

Module `Buffer` from the standard library allows to do concatenation in amortized linear time. Also, one could keep an index for the visited segment of `str`, instead of making a copy.